Improve Hamerly K-Means with OpenMP Implementation

MarkFischinger opened the pull request:
I've been working on optimizing the Hamerly K-means implementation, and I'm happy to share the results with you.

**Benchmark Results**
I ran a benchmark with the following settings:
```cpp
Number of data points: 100,000
Number of dimensions: 50
Number of clusters: 10
Maximum iterations: 100
```
**Here's what I found:**
```
Original version: 12.1609 seconds
OpenMP version: 6.98171 seconds
```

That's a 42.6% decrease in execution time!
Looking forward to your feedback! Let me know if you need any clarification or have any questions.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -132,6 +147,9 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
   double centroidMovement = 0.0;
+
+  #pragma omp parallel for reduction(+:distanceCalculations, centroidMovement) \
+                            reduction(max:furthestMovement, secondFurthestMovement)
```
What about `furthestMovingCluster`?  That is a little trickier to handle...

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -74,9 +79,13 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
     // First bound test.
     if (upperBounds(i) <= m)
     {
+      #pragma omp atomic
```
Does this need to be atomic if we already have the reduction for `hamerlyPruned`?

rcurtin reviewed (commented):
Nice improvement!  Just a couple small comments.

MarkFischinger commented:
I fixed the error and made sure to use `#ifdef MLPACK_USE_OPENMP` at the correct places. Here are the new benchmarks:
```console
mark@mark:~/gsoc/mlpack-benchmarks/src/kmeans$ ./hamerly_kmeans
Running benchmark...
Benchmark completed.
Total time: 4.95009 seconds
```

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -26,12 +26,12 @@ HamerlyKMeans<DistanceType, MatType>::HamerlyKMeans(const MatType& dataset,
 {
   // Nothing to do.
 }
-
```
No need to remove the empty line.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+        {
+          const double halfDist = dist / 2.0;
+          localMinClusterDistances(i) = std::min(localMinClusterDistances(i), halfDist);
+          localMinClusterDistances(j) = std::min(localMinClusterDistances(j), halfDist);
```
Can you explain the reasoning behind these changes? I don't umderstand the purpose or how this improves the implementation.  Unless I have missed something, this makes the algorithm significantly incorrect.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
       }
-      else if (dist < lowerBounds(i))
+
+      // The bounds failed. So test against all other clusters.
```
The original comment said significantly more than this for a reason. Please don't remove comments unless there is a very good reason.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
 
-  // Normalize centroids and calculate cluster movement (contains parts of
-  // Move-Centers() and Update-Bounds()).
+  // Normalize centroids and calculate cluster movement
```
There is no reason to change this comment.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
     // Calculate movement.
-    const double movement = distance.Evaluate(centroids.col(c),
-                                              newCentroids.col(c));
+    const double movement = std::sqrt(arma::sum(arma::square(centroids.col(c) - newCentroids.col(c))));
```
This is incorrect for any distance metric that is not the Euclidean distance. Please revert to the original implementation (even if it is slower due to the issue with `arma::norm()` that was discovered in the naive k-means PR.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -44,105 +44,146 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+    const double movement = std::sqrt(arma::sum(arma::square(centroids.col(c) - newCentroids.col(c))));
     centroidMovements(c) = movement;
     centroidMovement += std::pow(movement, 2.0);
-    ++distanceCalculations;
```
This is there for a reason, maybe if you removed it because of a threading issue it should be a `std::atomic`?

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -156,21 +197,31 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
 
-  // Now update bounds (lines 3-8 of Update-Bounds()).
-  for (size_t i = 0; i < dataset.n_cols; ++i)
+  // Now update bounds
```
Revert the comment please. It is extremely important for maintenance to be able to connect the code to the parts of the original paper.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -156,21 +197,31 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+      #pragma omp critical
+      {
+        Log::Warn << "Invalid assignment for point " << i << std::endl;
+      }
```
But can this ever happen? I don't think that it can (at least it couldn't in the original imementation).

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -156,21 +197,31 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
 
   return std::sqrt(centroidMovement);
 }
-
```
No need to remove the empty line separating the end of the function from the closing of the namespace.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -156,21 +197,31 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
 } // namespace mlpack
 
-#endif
+#endif
```
No need to remove the newline from the end of the file.

rcurtin reviewed (commented):
I'm not yet 100% sure everything here is correct (I am out of time for now, will come back later), but I think the general idea works and the speedup is nice. I have some structure and cleanliness comments; let me know if I can clarify them. :+1:

MarkFischinger commented:
@rcurtin I simplified the `minClusterDistance` update logic, made sure to use openmp more optimized by, for example, moving the centroid normalization and movement calculation into a parallel loop and fixed the calculation issue pointed out by you.

Before the changes these were the execution time results: 
```
DatasetSize,Time
1000,0.00677065
10000,0.148302
100000,1.62585
1000000,17.6833
```

After:
```
DatasetSize,Time
1000,0.004081
10000,0.0428959
100000,1.04919
1000000,3.32927
```

tested with this [benchmark script](https://gist.github.com/MarkFischinger/32778e1cacfc5dc080d4e53610f09458).

I also created this graph:

![benchmark_comparison](https://github.com/user-attachments/assets/011a7cdd-4f45-4aa5-b8ae-765849b6389f)

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
-
-    // First bound test.
-    if (upperBounds(i) <= m)
+    arma::mat threadNewCentroids(centroids.n_rows, centroids.n_cols,
+      arma::fill::zeros);
```
```suggestion
    arma::mat threadNewCentroids(centroids.n_rows, centroids.n_cols);
```
(Armadillo defaults to `fill::zeros`, so we don't need to include it here.)

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
-    if (upperBounds(i) <= m)
+    arma::mat threadNewCentroids(centroids.n_rows, centroids.n_cols,
+      arma::fill::zeros);
+    arma::Col<size_t> threadCounts(centroids.n_cols, arma::fill::zeros);
```
```suggestion
    arma::Col<size_t> threadCounts(centroids.n_cols);
```

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
   }
 
-  for (size_t i = 0; i < dataset.n_cols; ++i)
+  #pragma omp parallel
```
I was thinking to myself, "wouldn't it be nice if we could just use `#pragma omp parallel for reduction(+: centroids)` or something like this?"  That would be much nicer code-wise than making the `threadNewCentroids` object and `threadCounts` objects and then updating them.  Then I read about OpenMP custom reductions:

https://stackoverflow.com/questions/59034114/custom-omp-reduction-on-stdmaps

Do you want to give that a try and see if it works?  Don't feel obligated, we could also do it in another PR or another time.  But it could be nice for simplifying not just this code but other code that uses similar patterns for the other k-means strategies.  I guess that we could define Armadillo-specific custom reductions in some core header file, and then just use it where relevant.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+      arma::fill::zeros);
+    arma::Col<size_t> threadCounts(centroids.n_cols, arma::fill::zeros);
+    size_t threadHamerlyPruned = 0;
+    size_t threadDistanceCalculations = 0;
```
I think for these two variables you could directly use an OpenMP reduction?  Let me know if I overlooked something and there's a better reason to do it this way.

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+    size_t threadHamerlyPruned = 0;
+    size_t threadDistanceCalculations = 0;
+
+    #pragma omp for schedule(static)
```
Is `schedule(static)` the right thing here?  Much like Elkan's algorithm, we may be doing a different amount of work for each iteration.  Some iterations (or specifically some points) will be pruned away and exit early in the first or second bound tests, but some won't.  We can't statically know ahead of time which those will be, so it seems like a dynamic schedule would be the right one here.  I know it will be a little more overhead to do the dynamic scheduling, but I think it will pay off overall... what do you think?  This might be something that can only be validated empirically, but my intuition would suggest a dynamic schedule might be better (...at the same time, my intuition is often wrong!).

MarkFischinger replied:
I was unsure about changing it to dynamic, so I tested it and when set to static the performance was: 
```
Running benchmark for 1000 points...
Dataset size: 1000, Time: 0.00583865 seconds
Running benchmark for 10000 points...
Dataset size: 10000, Time: 0.0629136 seconds
Running benchmark for 100000 points...
Dataset size: 100000, Time: 0.383088 seconds
Running benchmark for 1000000 points...
Dataset size: 1000000, Time: 3.88048 seconds
Benchmark completed. Results saved to hamerly_kmeans_benchmark.csv
```

And when set to dynamic it resulted in: 
```
Running benchmark for 1000 points...
Dataset size: 1000, Time: 0.00631155 seconds
Running benchmark for 10000 points...
Dataset size: 10000, Time: 0.0597197 seconds
Running benchmark for 100000 points...
Dataset size: 100000, Time: 0.62253 seconds
Running benchmark for 1000000 points...
Dataset size: 1000000, Time: 6.61573 seconds
Benchmark completed. Results saved to hamerly_kmeans_benchmark.csv
```

rcurtin replied:
Ah, well, my intuition is wrong then.  Static it is!

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,70 +60,86 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
+
+        const double dist = distance.Evaluate(dataset.col(i), centroids.col(c));
+
+        // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i))
```
```suggestion
        // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i)).
```
The original period seems to have gotten lost 😄

MarkFischinger replied:
i removed it on purpose because it is above the 80 char limit, due to the indent, it is the 81st char. i thought removing it would be better than breaking the line and causing a larger diff size. let me know what is best here :)

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -132,6 +149,8 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
   size_t furthestMovingCluster = 0;
   arma::vec centroidMovements(centroids.n_cols);
   double centroidMovement = 0.0;
+  #pragma omp parallel for reduction(+:distanceCalculations,centroidMovement) \
+    schedule(static)
```
```suggestion
  #pragma omp parallel for reduction(+: distanceCalculations, centroidMovement) \
      schedule(static)
```
Just some small pedantic style fixes. :+1:

rcurtin reviewed (commented):
Awesome, thanks for the cleanups @MarkFischinger.  I think this one is basically ready to go but I had a thought about custom reductions that I want to see what you think about.  If you can address or respond to the handful of comments I've left, I think we can get this merged shortly. :+1:

MarkFischinger commented:
@rcurtin Reading the stackoverflow comment I implemented the idea of custom reductions. The code is now cleaner and has similar/slightly better performance:

```
Running benchmark for 1000 points...
Dataset size: 1000, Time: 0.00392151 seconds
Running benchmark for 10000 points...
Dataset size: 10000, Time: 0.0348763 seconds
Running benchmark for 100000 points...
Dataset size: 100000, Time: 0.352331 seconds
Running benchmark for 1000000 points...
Dataset size: 1000000, Time: 3.49524 seconds
Benchmark completed. Results saved to hamerly_kmeans_benchmark.csv
```

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -17,6 +17,14 @@
 
+// Custom reduction for arma::mat
+#pragma omp declare reduction(matAdd : arma::mat : omp_out += omp_in) \
+    initializer(omp_priv = arma::mat(omp_orig.n_rows, omp_orig.n_cols).zeros())
```
I think the `.zeros()` is unnecessary here since Armadillo will automatically initialize the matrix to zeros in recent versions.  (That should provide a little additional speedup too.)

rcurtin replied:
Also, some other little thoughts.  This is really nice, so it would be great if we could make it available to other techniques.  Can you move it to a new header file that we can put in, say, `mlpack/core/util/` and include in `mlpack/base.hpp`?  I guess you could call it `omp_reductions.hpp` or something like that.  I'm not totally tied to that so if you think there is a better place that works too.  But basically the idea is just that these reductions can be available for any mlpack technique.

rcurtin commented:
Awesome that the custom reduction worked!  That really simplifies the changes.  Do you think you would be up for opening a PR to make that change to use a custom reduction in the naive strategy too?  It should be a quick simplification (and I guess provide a slight amount of speedup too, given your results here).

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -118,20 +129,24 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
     ++counts(assignments[i]);
   }
 
+  distanceCalculations += threadDistanceCalculations;
```
If we're reducing `threadDistanceCalculations`, then couldn't we just use `distanceCalculations` directly?

rcurtin reviewed (commented):
Nice!  I think this is just about ready.  Do you want to add a note in `HISTORY.md` about the OpenMP k-means improvements?  You could reference this PR and the naive k-means PR, and then we could update it as we get the other PRs in. :+1:

I'm excited about getting this merged, the speedups for this one are really nice.  Once you handle the couple of comments from my side everything seems good to merge. :rocket:

rcurtin commented on HISTORY.md:
```diff
@@ -6,6 +6,12 @@ _????-??-??_
+ 
+ * Added OpenMP parallelization to Naive k-means (#3762).
+ 
+ * Added OpenMP parallelization to Elkan k-means (#3764).
```
Can we condense these into one line?

rcurtin commented on src/mlpack/core/util/omp_reductions.hpp:
```diff
@@ -0,0 +1,27 @@
+
+} // namespace mlpack
+
+#endif
```
Do you mind making sure this file has a newline at the end?

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -104,7 +107,7 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
       const double dist = distance.Evaluate(dataset.col(i), centroids.col(c));
 
-      // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i)).
+      // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i))
```
```suggestion
      // Is this a better cluster?  At this point, upperBounds[i] = d(i, c(i)).
```
Since we aren't adding an extra indentation, the period still fits. :+1:

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -82,7 +85,7 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
     // Tighten upper bound.
     upperBounds(i) = distance.Evaluate(dataset.col(i),
-                                       centroids.col(assignments[i]));
+                                      centroids.col(assignments[i]));
```
```suggestion
                                       centroids.col(assignments[i]));
```

rcurtin commented on src/mlpack/methods/kmeans/hamerly_kmeans_impl.hpp:
```diff
@@ -59,13 +60,15 @@ double HamerlyKMeans<DistanceType, MatType>::Iterate(const arma::mat& centroids,
     }
   }
 
+  size_t threadDistanceCalculations = 0;
+
```
```suggestion
```
This is unused now, so the variable isn't needed anymore.

rcurtin reviewed (approved):
Thanks for handling the issues!  It looks good to me now.  There are a couple other tiny comments I have, all very minor.  I can handle them before merge if you don't get to it, it's no big deal either way.
